src/test/niftypet_projector/test_ProjectorNiftyPET_adjoint.cxx

```diff
@@ -0,0 +1,446 @@
+        if (im.find_max() > 1e-4f)
+            break;
+        else 
+            std::cout << "\nRandom image contains all zeroes, regenerating...\n";
+    }
+}
+
+static
+void
+forward_project(ProjData &sino, const VoxelsOnCartesianGrid<float> &im, ForwardProjectorByBinNiftyPET &projector, double &time)
+{
+    std::cerr << "Forward projecting...\n";
+    sino.fill(0.f);
+
+    CPUTimer timer;
+    timer.start();
+
+    projector.set_input(im);
+    projector.forward_project(sino);
+
+    timer.stop();
+    time += timer.value();
+}
+
+static
+void
+back_project(VoxelsOnCartesianGrid<float> &im, const ProjData &sino, BackProjectorByBinNiftyPET &projector, double &time)
+{
+    std::cerr << "Back projecting...\n";
+    im.fill(0.f);
+
+    CPUTimer timer;
+    timer.start();
+
+    projector.start_accumulating_in_new_target();
+    projector.back_project(sino);
+    projector.get_output(im);
+
+    timer.stop();
+    time += timer.value();
+}
+
+static float get_inner_product(
+        const VoxelsOnCartesianGrid<float> &im1,
+        const VoxelsOnCartesianGrid<float> &im2)
+{
+    return std::inner_product(im1.begin_all(),im1.end_all(),im2.begin_all(),0.f);
+}
+
+static float get_inner_product(
+        const ProjDataInMemory &proj1,
+        const ProjDataInMemory &proj2)
+{
+    // Get number of elements
+    const size_t num_elements = proj1.size_all();
+
+    // Create arrays
+    std::vector<float> arr1(num_elements), arr2(num_elements);
+    proj1.copy_to(arr1.begin());
+    proj2.copy_to(arr2.begin());
+
+    return std::inner_product(arr1.begin(),arr1.end(),arr2.begin(),0.f);
+}
+
+static
+float
+test_inner_product(const VoxelsOnCartesianGrid<float> &im, const ProjData &sino,
+                   const VoxelsOnCartesianGrid<float> &im_proj, const ProjData &sino_proj)
+{
+    std::cerr << "Checking inner products...\n";
+    const float inner_product_images = get_inner_product(im,im_proj);
+    const float inner_product_sinos  = get_inner_product(sino, sino_proj);
+
+    std::cout << "\tinner product between images    = " << inner_product_images << "\n";
+    std::cout << "\tinner product between sinograms = " << inner_product_sinos << "\n";
+
+    if (std::abs(inner_product_images) + std::abs(inner_product_sinos) < 1e-4f) {
+        std::cout << "\n\t\tCan't perform adjoint test as both equal zero...\n";
+        std::cout << "\t\tmax in input image = " << im.find_max() << "\n";
+        std::cout << "\t\tmax in projected image = " << im_proj.find_max() << "\n";
+        std::cout << "\t\tmax in input image = " << find_max(sino) << "\n";
+        std::cout << "\t\tmax in projected image = " << find_max(sino_proj) << "\n";
+        return -1.f;
+    }
+
+    float adjoint_test =
+            std::abs(inner_product_images - inner_product_sinos) /
+            (0.5f * (std::abs(inner_product_images) + std::abs(inner_product_sinos)));
+    std::cout << "\t|<x, Ty> - <y, Tsx>| / 0.5*(|<x, Ty>|+|<y, Tsx>|) = " << adjoint_test << "\n";
```

Comment:
there should be some `check` call here. Are there any? Otherwise the test will always claim to succeed.